lib/libxdp/xsk.c

```diff
@@ -335,14 +356,25 @@ int xsk_umem__create_with_fd(struct xsk_umem **umem_ptr, int fd,
 	INIT_LIST_HEAD(&umem->ctx_list);
 	xsk_set_umem_config(&umem->config, usr_config);
 
-	memset(&mr, 0, sizeof(mr));
-	mr.addr = (uintptr_t)umem_area;
-	mr.len = size;
-	mr.chunk_size = umem->config.frame_size;
-	mr.headroom = umem->config.frame_headroom;
-	mr.flags = umem->config.flags;
+	if (umem->config.tx_metadata_len) {
+		memset(&mr, 0, sizeof(mr));
+		mr.addr = (uintptr_t)umem_area;
+		mr.len = size;
+		mr.chunk_size = umem->config.frame_size;
+		mr.headroom = umem->config.frame_headroom;
+		mr.flags = umem->config.flags;
+		mr.tx_metadata_len = umem->config.tx_metadata_len;
+		err = setsockopt(umem->fd, SOL_XDP, XDP_UMEM_REG, &mr, sizeof(mr));
+	} else {
+		memset(&mr_v1, 0, sizeof(mr_v1));
+		mr_v1.addr = (uintptr_t)umem_area;
+		mr_v1.len = size;
+		mr_v1.chunk_size = umem->config.frame_size;
+		mr_v1.headroom = umem->config.frame_headroom;
+		mr_v1.flags = umem->config.flags;
+		err = setsockopt(umem->fd, SOL_XDP, XDP_UMEM_REG, &mr_v1, sizeof(mr_v1));
```

Comment:
As an alternative to the _v1 struct, you could just use struct xdp_umem_reg from the kernel header, and pass the size as `offsetof(struct xdp_umem_reg, tx_metadata_len)` if tx_metadata_len is not set...